README
IceGenerator 0.5.5-pre1 - README

IceGenerator is a direct streaming generator for Icecast/Shoutcast. It runs as a daemon.
Three child processes handle all the stuff. A fourth (optional) child cares about
telnet communication.
Main idea for this program is radio automation on a Icecast/Shoutcast server. If you have
many server processes and CPU saving is your priority, u can use IceGenerator.
Obviously mp3s/oggs will be already resampled to allow their streaming on your bandwidth
capability.

This is a pre-realease version, cause many big changes are in progress. I'd like
that you can test it along its development.

============ FEATURES/LIMITATIONS ============
* Automatically put into playlist directories full of mp3s, or a MySQL table
* No resampling is allowed, so CPU is saved.
* MP3s/OGGs are played randomly or in alphabetic order, without repeating twice a song.
  You can loop the whole playlist at its end.
* When IceGenerator quits, finish its last song first.
* IceGenerator status is optionally reported to syslog/log file.
* Very useful when used with crond.
* ID3 support

============ REQUIREMENTS ============
* Linux 2.x/FreeBSD/NetBSD/MacOS/Solaris/OpenBSD.
* You have to install libshout2. You can find it at http://www.icecast.org/download.html.
* For ID3 support, you have to install id3lib. You can find it at http://id3lib.sourceforge.net
* An IceCast/Shoutcast server.
* An Intel 486-like CPU class (maybe a fast 386-like will work).
* A good broadband connection.


============ INSTALL ============
Just type:

./configure
make
make install

Configure options:
  --with-id3      adds ID3 support
  --with-mysql    adds MySQL support
  --with-pgsql    adds PostGreSQL support

Binary is installed in /usr/local/bin as default. 
It requires a configuration file, located in /usr/local/etc as default,
that indicates mp3s/oggs paths and all the stuff related to connection to Icecast.
A sample configuration file is in etc/ directory of this package.
All mp3s MUST HAVE .mp3 extension, and oggs MUST HAVE .ogg extension. Case is
irrilevant.

Configure your streaming server properly. The default user for Icecast-2.0 is "source",
but you can change as you want. You can also set a mount point different from "/", so you
an many instances of icegenerator at once (each one with a different mount point, obviously).

Then type "icegenerator" and put on your stream player (like XMMS). If you can't
hear music, maybe your volume is too low :-D

More options (like icegenerator.conf path, deamon start) are available
at command line. Type "icegeneator -h" for synopsys.

To quit this daemon, just run "killall icegenerator", and wait for last
song to finish.

Really simple, uh?

NOTE:
Some players (like WinAmp, XMMS, etc.) requires a valid mount point for OGG files. They
need .OGG extension, cause MIME header isn't evaluated. Keep in mind it ...


============ CONFIGURATION ============
All configuration options are listed in icegenerator.conf file that usually
resides in /usr/local/etc of your sistem.

Each line that starts with a '#' is treated as a comment.

*** Server section ***
All option in server section are required.

- IP: Specify here your IceCast/ShoutCast server IP
- PORT: Specify here your IceCast/ShoutCast server port
- SERVER: Specify what protocol your IceCast/ShoutCast uses. Refer to your
  stream server documentation for further details. Only two protocols are
  supported, ICY and HTTP.
- MOUNT: Specify mount point for clients, that is where they find your stream
  server in http path. For example, setting MOUNT as /my_stream requires that
  clients use URL http://your_server:your_port/my_stream. REMEMBER TO SET A VIRTUAL
  MOUNT POINT HERE FOR OGG STREAMING. MANY PLAYERS (LIKE WINAMP, XMMS, ETC.) CAN'T
  AUTO RECOGNIZE OGG STREAMING WITHOUT .OGG EXTENSION
 
*** User section ***
All option in user section are required.

- USER: Specify which user has streaming rights. As far as I know, IceCast uses
  "source" as default, but u can use anything else, cause it's ignored. However
  you can find more informations in your stream server documentation.
- PASSWORD: Specify password to connect to your stream server.

*** File and playlist section ***
All option in file and playlist section are required.

- FORMAT: Specify what type of audio data are you going to stream. Only MP3 and OGG
  files are allowed. Use "0" for OGG, "1" for MP3.
- MP3PATH: Here is one of strong point of IceGenerator, its capability to accept
  many ways to search for audio songs. First three chars of this variable specify
  where are your audio songs, use "PTH" for directories, "SQL" for a MySQL table
  (only allowed if MySQL support is enabled during compilation), "PQL" for a PostGreSQL
  table (only allowed if PostGreSQL support is enabled during compilation), "PLS" for .PLS
  playlist file support, "M3U" for .M3U playlist file support. After a ':' character,
  you can specify all option that refers to way that you have chosen.
  + PTH case: you can specify one or more directories separated with ';'. Wrong paths
    will be skipped.
  + PQL/SQL case: you have to specify these informations separated with ';' in the
    following order:
    % host: your SQL server ip or symbolic name.
	  % port: your SQL server port. If you specify "0", default port will be used.
	  % user: SQL user that has access rights to database that you'll specify. Only
	    SELECT right is required.
	  % password: password for SQL user.
	  % database: database name that contains table that contains audio songs paths.
		% table: table name that contains audio songs paths.
		% field: field name that contains audio songs paths.
  + M3U/PLS case: you have to specify only a playlist file path.
		
	Examples:
	+ You have two directories full of OGG files, /path1/dir and /my_audio/rock/80s.
	  MP3PATH=pth:/path1/dir;/my_audio/rock/80s
	
	+ You have a MySQL server located at sql.mydomain.com, you have a database called
	  webextra with a table called songsdata and a column called songs. MySQL user
		that has access rights is called john.
		MP3PATH=sql:sql.mydomain.com;0;john;secretpassword;webextra;songsdata;songs

- RECURSIVE: When this variable is set to 1, IceGenerator search for audio songs
  in subdirectories of directories specified in MP3PATH. Obviously, it is ignored
	when you use MySQL as source data.
	
- DUMPFILE: Specify a file where stream is saved. 

- LOOP: If set, playlist is looped at its end.

- SHUFFLE: If set, playlist is played randomly.

*** Information section ***
- NAME:
- GENRE:
- DESCRIPTION:
- URL: Specify name, genre, description and URL of your stream.

- BITRATE: Specify bitrate of your stream. This value is purely informational,
  it doesn't affect real bitrate.
	
- PUBLIC: If set, your URL is published into stream directory server(s).

- METAUPDATE: Specify delay (in seconds) between META information update that
  is sent to clients. A value of "0" disables META information change, and
  only NAME value is displayed.

- MDFPATH: Specify path of your global MetaDataFile. See README.IceMetal
  for more info.
	
*** Debug and logging information ***
- LOG: Specify logging type. "0" means no logging at all, "1" logs to system
  log daemon, "2" logs to /var/log/icegenerator.log.
	
*** Telnet interface ***
- DATAPORT: Specify port used for telnet interface. If not specified at all,
  telnet interface will be disabled. A value of "0" sets default port (8796).
	

============ TELNET INTERFACE ============
You can connect to IceGenerator during stream with a TCP connection like telnet.
Connection port is specified with DATAPORT option. After command is sent, an
answer is returned and connection is closed. Commands that not require an answer
have an "OK" string in return. All commands MUST HAVE a '.' character at their end.
Available commands are:
- TRACK: returns information about current track.
- LOOP: returns "1" or "0" if loop is active or not.
- SHUFFLE: returns "1" or "0" if shuffle is active or not.
- LOOPON: activates loop.
- LOOPOFF: deactivates loop.
- SHUFFLEON: activates shuffle.
- SHUFFLEOFF: deactivates shuffle.
- QUIT: quit after current track.
- QUITNOW: quit immediately.

More commands will be available in future.

============ ID3 AND ICEMETAL ============
See README.IceMetal for info.


============ ACKNOWLEDGEMENTS ============
I'd to thank

  - DIeMONd
  - TSx7 (alias techtonik)
  - Stefani Banerian
  - Michel Hendriks
  - Jacob Brandup
  - Patrick Atlas
  
for help me fixing bugs and suggesting more features.

Many thanks also to SourceForge, FreshMeat, Berlios, SourceWell, IceWalkers for
hosting and promoting this project.


============ CONTACT ============
For all suggestion and bug reporting, write to administrator@tortugalabs.it.


GOOD LISTENING!!!!!!!


ChangeLog
Version 0.5.5-pre2:
  - FIXED: Serious configure script bug

Version 0.5.5-pre1:
  - FIXED: Serious bug in configuration file parser module
  - FIXED: Don't try to open non-existing files during stream
  - FIXED: Race condition on meta streamer <-> selector launch
  - FIXED: Memory leaks
  - Added ID3 support
  - Added ICEMETAL language support
  - Default icegenerator.conf path is now SYSCONFDIR-fixed

Version 0.5.4:
  - FIXED: Incorrect socket definitions on MacOSX
  - FIXED: Segfault occurs if log object is NULL when error
  - FIXED: --with-shout parameter (-lshout -lvorbis missing)
  - Added PostGreSQL support
  - Added .M3U playlist support
  - Added .PLS playlist support
  - Removed basename function cause is not POSIX compliant
  - Removed no-rtti option, it causes segfaults on some systems
  - Log file is now path configurable
  - Added more info into README file

Version 0.5.3:
  - FIXED: Missing declaration on Mac OS X
	- FIXED: Alpha signal.h compilation error
  - FIXED: removed POSIX definition, cause BSD conflict

Version 0.5.2:
  - Added telnet minimal interface
  - Added MySQL support
	- Code cleaner (again)
	- FIXED: heading '/' in log file path
	- FIXED: massive bugfix in configure script
	
Version 0.5.1:
  - Code cleaner
  - Added SOURCE option
  - Added LOOP option
  - Added SHUFFLE option
  - Added METAUPDATE option
  - FIXED: Serious bug in SERVER option
  - FIXED: try/catch mechanism didn't work properly
  - FIXED: possible error in parsing configuration
  - FIXED: wrong path bug
  - FIXED: opendir possibly crash (thanks to DIeMONd)
  - FIXED: multiple directories existing check
  - Songs are sorted by filename on SHUFFLE = 0
  - Now IceG is a GNU standard package, hope that multiple platforms work now
  - Added MP3/OGG recursive directory search option (thx to DIeMONd again for suggesting me)
  - MP3/OGG file extension search is now case insensitive
  - Update time for metadata information is configurable, even metadata information presence too

Version 0.5:
  - Libshout 1.3.x no longer supported
  - Full POSIX implementation
  - Multiple playlist directories
  - Support for OGG Vorbis
  - Command line interface
  - Multiple log configuration
  - Threads (no more detached processes)
  - All "external errors" are catched
  - Clean exit, all system resources are released
  - FreeBSD/NetBSD/MacOS/Linux 2.x platform supported

Version 0.3:
  - Added libshout2 support.
  - Many bugfixes
  - Stronger configuration file parser
  - Support for Icecast-1.3.21/Icecast-2.0/Shoutcast server
  - Removed ICY_COMPAT option cause is useless

Version 0.1:
  - Initial release


src/main.cpp
/*
 *  main.cpp -- Simple daemon to source an icecast server with direct
 *              mp3/ogg streaming, no resampling.
 *
 *  Copyright (C) 2004 Tony Sin(x) '76 <administrator@tortugalabs.it>
 *  All rights reserved.
 *
 */

/*
 *			 GNU GENERAL PUBLIC LICENSE
 *			    Version 2, June 1991
 *
 *  Copyright (C) 1989, 1991 Free Software Foundation, Inc.
 *                           675 Mass Ave, Cambridge, MA 02139, USA
 *  Everyone is permitted to copy and distribute verbatim copies
 *  of this license document, but changing it is not allowed.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.

 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.

 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
 
/* Version history:
   0.1 - First release
   0.3 - Libshout 2.0 support
         Strong configuration parses
         Support for Icecast2.0/Shoutcast
         Bug fixes
   0.5 - Libshout 1.3.x no longer supported
         Full POSIX implementation
         Multiple playlist directories
         Support for OGG Vorbis
         Command line interface
         Multiple log configuration
         Threads (no more detached processes)
         All "external errors" are catched
         Clean exit, all system resources are released
         FreeBSD/NetBSD/Linux 2.x/MacOS platform supported
 0.5.1 - Code cleaner
         Added SOURCE option
         Added LOOP option
         Added SHUFFLE option
         Added METAUPDATE option
         FIXED: Serious bug in SERVER option
         FIXED: try/catch mechanism didn't work properly
         FIXED: possible error in parsing configuration
         FIXED: wrong path bug
         FIXED: opendir possibly crash (thanks to DIeMONd)
         FIXED: multiple directories existing check
         Songs are sorted by filename on SHUFFLE = 0
         Now IceG is a GNU standard package, hope that multiple platforms work now
         Added MP3/OGG recursive directory search option (thx to DIeMONd again for suggesting me)
         MP3/OGG file extension search is now case insensitive
         Update time for metadata information is configurable, even metadata information presence too
 0.5.2 - Added telnet minimal interface
         Added MySQL support
         Code cleaner (again)
         FIXED: heading '/' in log file path
         FIXED: massive bugfix in configure script
 0.5.3 - FIXED: Missing declaration on Mac OS X
         FIXED: Alpha signal.h compilation error
         FIXED: removed POSIX definition, cause BSD conflict
 0.5.4 - FIXED: Incorrect socket definitions on MacOSX
         FIXED: Segfault occurs if log object is NULL when error
         FIXED: --with-shout parameter (-lshout -lvorbis missing)
         Added PostGreSQL support
         Added .M3U playlist support
         Added .PLS playlist support
         Removed basename function cause is not POSIX compliant
         Log file is now path configurable
         Added more info into README file
 0.5.5 - FIXED: Serious bug in configuration file parser module
         FIXED: Don't try to open non-existing files during stream
         FIXED: Race condition on meta streamer <-> selector launch
         FIXED: Memory leaks
         Added ID3 support
         Added ICEMETAL language support
         Default icegenerator.conf path is now SYSCONFDIR-fixed
*/

#ifdef HAVE_CONFIG_H
  #ifndef __main_config_h__
    #define __main_config_h__
    #include "../config.h"
  #endif

  #ifndef __GNU_LIBRARY__
    #define __GNU_LIBRARY__
  #endif

  #ifdef HAVE_GETOPT_H
    #include <getopt.h>
  #endif

  #ifdef HAVE_STDIO_H
    #include <stdio.h>
  #endif
 
  #ifdef HAVE_STDLIB_H
    #include <stdlib.h>
  #endif

  #ifdef HAVE_PTHREAD_H
    #include <pthread.h>
  #endif

  #ifdef HAVE_SYS_TYPES_H
    #include <sys/types.h>
  #endif

  #ifdef HAVE_SYS_STAT_H
    #include <sys/stat.h>
  #endif
  
  #ifdef HAVE_FCNTL_H
    #include <fcntl.h>
  #endif

  #ifdef HAVE_CTYPE_H
    #include <ctype.h>
  #endif
	
  #ifdef HAVE_STRING_H
    #include <string.h>
  #endif

  #ifdef HAVE_UNISTD_H
    #include <unistd.h>
  #endif
  
  #define APPNAME PACKAGE_STRING
#else
  #define APPNAME "IceGenerator"
#endif

#include "log.h"
#include "config.h"
#include "sgnl_handler.h"
#include "streamer.h"
#include "circular.h"
#include "synch.h"
#include "main.h"
#include "utils.h"

#define HELP_MSG_LINES 8

const char *help_msg[HELP_MSG_LINES] = { APPNAME,
                                         NULL,
                                         "Usage:",
                                         "\ticegenerator [-h] [-d] [-f filename]",
                                         NULL,
                                         "\t-h\t\tDisplay this help message",
                                         "\t-f filename\tSpecify configuration file",
                                         "\t-d\t\tRun in foreground - no daemon" };
cLog *log_obj = NULL;
cSignal *sig_obj = NULL;
cConfig *conf_obj = NULL;
cStreamer *stream_obj = NULL;
cCircularList *ready_obj = NULL, *played_obj = NULL;
cSemaphore *sem_obj = NULL;
int track_load, end_track, data_mutex, start_meta;
volatile bool time_to_quit = false;
volatile bool quit_now = false;
bool ok_to_end = false, changed = false;
char ErrStr[127] = "Unknown error";

void cmd_quit() {
  time_to_quit = true;
}

void cmd_quit_now() {
  time_to_quit = true;
  quit_now = true;
}

enum signal_action_type {
  signal_action_ignore,
  signal_action_quit,
  signal_action_quit_now,
  signal_action_quitnow = signal_action_quit_now
};

signal_action_type signal_action = signal_action_quit;

static void signal_termination_proc(const int sig)
{
  switch (signal_action) {
    case signal_action_ignore:
      /* do nothing */
      break;
    case signal_action_quit:
    {
      char * signal_name = m_strsignal2(sig);
      log_obj->printf("Got signal %s. Waiting for all child processes to terminate...", signal_name);
      free(signal_name);
      cmd_quit();
      break;
    }
    case signal_action_quit_now:
    {
      char * signal_name = m_strsignal2(sig);
      log_obj->printf("Got signal %s. Quitting now...", signal_name);
      free(signal_name);
      cmd_quit_now();
      break;
    }
  }
}

void clean_objs()
{
  if (sem_obj != NULL)
    delete sem_obj;

  if (played_obj != NULL)
    delete played_obj;
  
  if (ready_obj != NULL)  
    delete ready_obj;
  
  if (stream_obj != NULL)
    delete stream_obj;
  
  if (conf_obj != NULL)
    delete conf_obj;
  
  if (sig_obj != NULL)
    delete sig_obj;
  
  if (log_obj != NULL)
    delete log_obj;
}

int
#ifdef INSIDE_CATCH
programMain
#else
main
#endif
(int argc, char **argv)
{
  int i, c;
  bool is_daemon = true;
  pthread_t player_tid, selector_tid, data_streamer_tid, data_server_tid;

  try
  {
    /* Parse command line parameters and reads configuration file */
    if (argc > 1)
      while ((c = getopt (argc, argv, "hdf:")) != -1)
        switch (c)
        {
      	  case 'h': for (i = 0; i < HELP_MSG_LINES; i++)
      	            {
      	              if (help_msg[i] != NULL)
      	                printf("%s",help_msg[i]);
      	              printf("\n");
      	            }
      	            printf("\n");
      	            exit(0);
      	            break;
      	  case 'f': conf_obj = new cConfig(optarg);
      	                       break;
      	  case 'd': is_daemon = false;
      	            break;
      	  case ':': printf("Invalid parameter\n");
      	            exit(1);
      	            break;
        }
    
    if (conf_obj == NULL)
      conf_obj = new cConfig(DEFAULT_CONF_FILE);

    {
      const char * action = conf_obj->GetValue(CONFIG_SIGNAL_ACTION);
      if (!action)
        action = "quit";
      #define CHECK_ACTION(a) \
        else if (strcasecmp(action, #a) == 0) { \
          signal_action = signal_action_##a; \
        }
      if (0) {}
      CHECK_ACTION(ignore)
      CHECK_ACTION(quit)
      CHECK_ACTION(quitnow)
      CHECK_ACTION(quit_now)
      #undef CHECK_ACTION
    }

    /* Init log file */
    log_obj = new cLog(conf_obj);

    /* Select streaming type and scans directories */
    ready_obj = new cCircularList;
    ready_obj->ParsePlayList(conf_obj);
    if (ready_obj->Empty())
    {
      strcpy(ErrStr,"No files found");
      throw 0;
    }
    else
    {
      played_obj = new cCircularList;
      ready_obj->SetStart();
    }

    /* Opens connection to stream server */
    stream_obj = new cStreamer;
    stream_obj->SetData(conf_obj);
    stream_obj->Connect();

    log_obj->WriteLog("Connected to stream server");
    
    /* Init synchronization module */
    sem_obj = new cSemaphore;
    data_mutex = sem_obj->AddSem(1);
    track_load = sem_obj->AddSem(0);
    end_track = sem_obj->AddSem(0);
    start_meta = sem_obj->AddSem(0);

    /* Launches as a daemon */
    if (is_daemon)
    {
      fclose(stdin);
      fclose(stdout);
      fclose(stderr);
      if (fork())
        exit(0);
      setsid();
      if (fork())
        exit(0);
      chdir("/");
      umask(0);
      log_obj->WriteLog("Going to daemon land...");
    }

    /* Install signal handler */
    sig_obj = new cSignal();
    sig_obj->AddHandler(SIGHUP,signal_termination_proc);
    sig_obj->AddHandler(SIGTERM,signal_termination_proc);
    sig_obj->AddHandler(SIGINT,signal_termination_proc);
    sig_obj->AddHandler(SIGQUIT,signal_termination_proc);
    sig_obj->AddHandler(SIGABRT,signal_termination_proc);

    /* Generates threads */
    pthread_create(&player_tid,NULL,player,NULL);
    pthread_create(&data_streamer_tid,NULL,data_streamer,NULL);
    pthread_create(&selector_tid,NULL,selector,NULL);
    pthread_create(&data_server_tid,NULL,data_server,NULL);

    /* Doesn't care about SIGS */
    sig_obj->Block();

    /* Waits for childen to die */
    pthread_join(player_tid,NULL);
    pthread_join(data_streamer_tid,NULL);
    pthread_join(selector_tid,NULL);
	
    if (conf_obj->GetValue(CONFIG_DATAPORT) != NULL)
    {
      close_data_server();
      pthread_join(data_server_tid,NULL);
    }

    log_obj->WriteLog("Have a nice day");

    /* Time to go out */
    stream_obj->Disconnect();
    clean_objs();
  }
  catch (...)
  {
    /* Something went wrong ... */
    if (is_daemon && (log_obj != NULL))
      log_obj->WriteLog(ErrStr);
    else
      printf("%s\n\n",ErrStr);

    clean_objs();
  }

  return 0;
}
